src/engine/LocalVocab.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright 2022, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Author: Hannah Bast <bast@cs.uni-freiburg.de>
+
+#include "engine/LocalVocab.h"
+
+#include "absl/strings/str_cat.h"
+#include "global/Id.h"
+#include "global/ValueId.h"
+
+// _____________________________________________________________________________
+Id LocalVocab::getIdAndAddIfNotContained(const std::string& word) {
+  // The following code avoids computing the hash for `word` twice in case we
+  // see it for the first time (note that hashing a string is not cheap). The
+  // return value of the `insert` operation is a pair, where `result.first` is
+  // an iterator to the (already existing or newly inserted) key-value pair, and
+  // `result.second` is a `bool`, which is `true` if and only if the value was
+  // newly inserted.
+  auto [keyValuePair, isNewWord] = wordsToIdsMap_.insert({word, nextFreeId_});
+  if (isNewWord) {
+    idsToWordsMap_.push_back(&(keyValuePair->first));
+    nextFreeId_ = Id::makeFromLocalVocabIndex(
+        LocalVocabIndex::make(idsToWordsMap_.size()));
+  }
```

Comment:
I am not sure if storing the next id is really worth the hazzle.
Maybe just calculate it at the beginning of this function.
The `size()` function is a cheap lookup that is always hot in cache/registers,
`make` is a noop, and `makeFromLocalVocabIndex` is 1 or 2 cycles (bitwise operations with masks). 
But it is your call, your implementation is faster, but I slightly doubt that the difference is measurable:)